aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +22,333 @@ namespace at {
-          ATENSOR##_dim++; \
-      } \
-      /* Allocate an array of 3*dim elements, where dim is the number of contiguous sections */ \
-      ATENSOR##_counter = new int64_t[3*ATENSOR##_dim]; \
-      ATENSOR##_sizes = ATENSOR##_counter + ATENSOR##_dim; \
-      ATENSOR##_strides = ATENSOR##_counter + 2*ATENSOR##_dim; \
-      TH_TENSOR_dim_index = ATENSOR##_dim-1; \
-      ATENSOR##_dimOffset = (DIM == ATENSOR.dim()-1) ? &ATENSOR##_i : &ATENSOR##_counter[DIM]; \
-      ATENSOR##_sizes[TH_TENSOR_dim_index] = ATENSOR.sizes()[ATENSOR.dim()-1]; \
-      ATENSOR##_strides[TH_TENSOR_dim_index] = ATENSOR.strides()[ATENSOR.dim()-1]; \
-      /* ATENSOR##_counter tracks where we are in the storage. The offset into the */ \
-      /* storage is given by storage_offset + (i * j), where i is the stride */ \
-      /* vector and j is tensor_counter vector. This sets the starting position for the loop. */ \
-      for(ATENSOR##_i = ATENSOR##_dim-1; ATENSOR##_i >= 0; --ATENSOR##_i) { \
-        ATENSOR##_counter[ATENSOR##_i] = 0; \
-      } \
-      for(ATENSOR##_i = ATENSOR.dim()-2; ATENSOR##_i >= 0; --ATENSOR##_i) { \
-        if (ATENSOR.strides()[ATENSOR##_i] == ATENSOR.strides()[ATENSOR##_i+1] * ATENSOR.sizes()[ATENSOR##_i+1] && ATENSOR##_i != DIM && ATENSOR##_i+1 != DIM) { \
-          ATENSOR##_sizes[TH_TENSOR_dim_index] = ATENSOR.sizes()[ATENSOR##_i] * ATENSOR##_sizes[TH_TENSOR_dim_index]; \
-          if (DIM != ATENSOR.dim()-1 && ATENSOR##_i < DIM) \
-            ATENSOR##_dimOffset--; \
-        } else { \
-          --TH_TENSOR_dim_index; \
-          ATENSOR##_sizes[TH_TENSOR_dim_index] = ATENSOR.sizes()[ATENSOR##_i]; \
-          ATENSOR##_strides[TH_TENSOR_dim_index] = ATENSOR.strides()[ATENSOR##_i]; \
-        } \
-      } \
-      /* Size of the inner most section */ \
-      ATENSOR##_size = ATENSOR##_sizes[ATENSOR##_dim-1]; \
-      /* Stride of the inner most section */ \
-      ATENSOR##_stride = ATENSOR##_strides[ATENSOR##_dim-1]; \
-    } \
-  } \
-  ATENSOR##_i = 0;
+template <typename T>
+struct util_tensor {
+  T* data = NULL;
+  int64_t *counter = NULL, *sizes = NULL, *strides = NULL, *dimOffset = NULL;
+  int64_t stride = 0, size = 0, dim = 0, i;
+  int contiguous;
+  util_tensor(Tensor& tensor, int64_t dim, bool ALLOW_CONTIGUOUS) {
+    dim = dim;
+    int64_t TH_TENSOR_dim_index = 0;
+    int contiguous = ALLOW_CONTIGUOUS && dim < 0;
+    data = tensor.data<T>();
+    size = 1;
+    stride = 1;
+    for (i = tensor.dim() - 1; i >= 0; i--) {
+      if (tensor.sizes()[i] != 1) {
+        if (tensor.strides()[i] == size && i != dim)
+          size *= tensor.sizes()[i];
+        else {
+          contiguous = 0;
+          break;
+        }
+      }
+    }
+    if (!contiguous) {
+      dim = 1;
+      for (i = tensor.dim() - 2; i >= 0; i--) {
+        if (tensor.strides()[i] !=
+                tensor.strides()[i + 1] * tensor.sizes()[i + 1] ||
+            i == dim || i + 1 == dim)
```

Comment:
This is really hard to read. Can you break out at least the first part into `bool something == ...`, and use this in the condition?